### PR TITLE
Allow custom application of pageParam and sanitization of key

### LINKED ISSUE #49: Remove pageParamKey from queryKey in useInfiniteQuery
Related to #43 there is some potential confusion because we encode the contents of the incoming message into the key of the query, including pagination information which changes actively due to change of pagination.

Omitting `pageParamKey` may be sufficient but also may not cover everything (like if the page param is deeply nested in the message).

### PR DESCRIPTION
**Pull request by paul-sachs**
Allow customizing the stored query key for paginated queries, as well as more flexible ways of applying said page param.

Fixes #49

### CODE REVIEW

File: packages/connect-query/src/unary-hooks.ts [hunk 1]
```diff
@@ -106,14 +106,40 @@ export interface UnaryHooks<I extends Message<I>, O extends Message<O>> {
...
+          ) => PartialMessage<I>[ParamKey];
+          /**
+           * The option allows you to remove fields or otherwise customize the input used to generate the query key.
+           * By default, we will remove the pageParamKey from the input. If this is provided, we will use this result instead.
+           */
+          sanitizeInputKey?: (input: PartialMessage<I>) => unknown;
+          onError?: (error: ConnectError) => void;
+          transport?: Transport | undefined;
+          callOptions?: CallOptions | undefined;
+        },
```
Comment (paul-sachs) [inline packages/connect-query/src/unary-hooks.ts:142]: There is probably a better way to do this, but moving it to a shared interface makes it so type assertions on the param result seems to break without exposing those shared types.

File: packages/connect-query/src/unary-hooks.ts [hunk 2]
```diff
@@ -106,14 +122,18 @@ export interface UnaryHooks<I extends Message<I>, O extends Message<O>> {
...
-    },
+    options: BaseInfiniteQueryOptions<I, O, ParamKey> &
+      (
+        | {
+            applyPageParam: (options: {
+              pageParam: PartialMessage<I>[ParamKey] | undefined;
+              input: PartialMessage<I>;
+            }) => PartialMessage<I>;
+          }
+        | {
```
[the remark this answers is not in the document]
Comment (paul-sachs) [reply packages/connect-query/src/unary-hooks.ts:133]: Hmm, yeah. I guess we'll need something like that. Not sure how I would do function overloads within a single interface.

[the remark this answers is not in the document]
Comment (dimitropoulos) [reply packages/connect-query/src/unary-hooks.ts:133]: as an aside: function overloads are just intersections.  if you need to create one inline, you can do it like this:
Screenshot 2023-05-30 at 14 03 06

[the remark this answers is not in the document]
Comment (paul-sachs) [reply packages/connect-query/src/unary-hooks.ts:133]: hmm interesting. Yeah i might be able to use that. Let me take a look.

**Review by smaye81 — APPROVED** [review]

**Pull request merged**